java/core/src/main/java/com/microsoft/bond/BondType.java

```diff
@@ -0,0 +1,514 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.bond;
+
+import com.microsoft.bond.helpers.ArgumentHelper;
+import com.microsoft.bond.protocol.ProtocolWriter;
+import com.microsoft.bond.protocol.TaggedProtocolReader;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Represents a type in the Bond type system and implements behavior specific to that type, such as
+ * schema inference, initialization, serialization and deserialization. A Bond type cane be a primitive,
+ * a struct, a specialization of a generic type (including containers, nullable, or generic Bond structs).
+ * @param <T> the class of the value
+ */
+public abstract class BondType<T> {
+
+    // The global type cache
+    static final TypeCache typeCache = new TypeCache();
+
+    // Registry of struct type resolvers; entries are added by static class initializers
+    private static final ConcurrentHashMap<
+            Class<? extends BondSerializable>,
+            StructBondTypeResolver<? extends BondSerializable>> structTypeResolverRegistry =
+            new ConcurrentHashMap<
+                    Class<? extends BondSerializable>,
+                    StructBondTypeResolver<? extends BondSerializable>>();
+
+    // Registry of enum types; entries are added by static class initializers
+    private static final ConcurrentHashMap<
+            Class<? extends BondEnum>,
+            EnumBondType<? extends BondEnum>> enumTypeRegistry =
+            new ConcurrentHashMap<
+                    Class<? extends BondEnum>,
+                    EnumBondType<? extends BondEnum>>();
+
+    // precomputed hash code helps to speed up object retrieval from the type cache
+    private final int precomputedHashCode;
```

Comment:
It's a useful optimization since we heavily rely on type caching. Precomputed hash code also helps object equality (if objects hash codes don't match then no need to do any deep object comparison).

---
In reply to: [123400869](https://github.com/Microsoft/bond/pull/502#discussion_r123400869) [](ancestors = 123400869)